### PR TITLE
Allow configuring pagination limit in settings

### LINKED ISSUE #483: Images per page command line option
I want to see over 12 images on single view.
So I want --images-per-page=num_images command line option.

### PR DESCRIPTION
**Pull request by wchargin**
Summary:
In accordance with #483, this patch adds a configuration option to allow
changing the number of visualizations per page to a value other than 12.

Closes #483.

Test Plan:
Load the histogram dashboard with the histogram demo, which has seven
visualizations. Open the settings menu, and note that the default
pagination limit value is 12. Change the value to 1, close the dialog,
and move to page 6. Then, reopen the dialog and change the pagination
limit to 2; note that you are now on page 3 of 4. (This tests that the
existing-but-never-before-used behavior of `tf-pagination-view` that
attempts to retain approximately the same view across pagination limit
changes still works.)

Try to change the value to 0 or a negative value; note that this fails
without any console errors.

wchargin-branch: configurable-pagination-limit

### CODE REVIEW

File: tensorboard/components/tf_paginated_view/tf-paginated-view.html [hunk 1]
```diff
@@ -237,13 +237,16 @@
...
+    attached() {
+      this._limitListener = () => {
+        this.set('_limit', store.getLimit());
+      };
+      store.addLimitListener(this._limitListener);
+      this._limitListener();
+    },
+
+    detached() {
+      store.removeLimitListener(this._limitListener);
```
Comment (chihuahua) [inline tensorboard/components/tf_paginated_view/tf-paginated-view.html]: Nice performance improvement for people away from the tab.

File: tensorboard/components/tf_paginated_view/paginatedViewStore.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ * Remove a listener registered with `addLimitListener`.
+ */
+export function removeLimitListener(listener: Listener): void {
+  listeners.delete(listener);
+}
+
+export function getLimit() {
+  if (_limit == null) {
+    _limit = getNumber(LIMIT_LOCAL_STORAGE_KEY, /*useLocalStorage=*/true);
+    if (_limit !== +_limit || _limit <= 0) {
```
[the remark this answers is not in the document]
Comment (wchargin) [reply tensorboard/components/tf_paginated_view/paginatedViewStore.ts]: I was trying to detect things that weren't actually numbers, to handle the case where (e.g.) the user has manually changed their `localStorage` value to `"hello"`, or it's gotten corrupted somehow.

But I now realize that this isn't necessary: `getNumber` will always return either `undefined` or a number. So I can just use `_limit == null || _limit <= 0`.

File: tensorboard/components/tf_paginated_view/tf-paginated-view.html (hunk 1 above)
Comment (wchargin) [reply tensorboard/components/tf_paginated_view/tf-paginated-view.html]: I believe that it's also necessary for correctness: otherwise, paginated views that get unmounted will try to `set` polymer properties on instances that no longer exist, which should be undefined behavior.

Comment (chihuahua) [reply tensorboard/components/tf_paginated_view/tf-paginated-view.html]: Ah k! Makes sense.

File: tensorboard/components/tf_paginated_view/paginatedViewStore.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (chihuahua) [reply tensorboard/components/tf_paginated_view/paginatedViewStore.ts]: LGTM

**Comment by dandelionmane** [comment]
First, nice work!

Second: This is a pretty complicated pattern (defining new listeners for the particular setting, registering listeners, etc). Will we need to duplicate this machinery every time we add a new global setting? Wondering if we can factor out simpler shared machinery for it. 

It would be nice if we could make it very easy for every plugin to save lots of settings in localstorage.

There's also a case for putting this info into the URL - it would be nice if copy+pasting someone else's tensorboard URL would give you the same view they have (otherwise, I send you my URL and say "look at the 13th chart here" and you get confused because your TB displays only 12 charts. A bit contrived in this case). Although it would be weird if loading someone else's URL wiped your TB settings.

On a high level, we might be better off moving settings into the TensorBoard server rather than the frontend, because then it can persist for multiple people using the same TB (feature or bug?) and survive migrations where the URL changes (which would wipe the localStorage).

tl;dr: Figuring out the right way to do settings in TB is confusing, and we have a few different approaches in flight.

Either way, I'm happy with this PR as-is. Mostly just bringing up future considerations.

**Comment by wchargin** [comment]
Thanks!

> pretty complicated pattern

I’m inclined to agree: if not _complicated_ per se, it’s certainly verbose and error-prone. We can definitely factor out shared machinery. The main reason that I didn’t do so prior to this commit is just that, as you mention, it’s not yet clear what the right abstraction is, and the wrong abstraction is worse than no abstraction.

Just to put this out there—I don’t think that we need a Flux-like or Redux-like architecture at this time. (But perhaps I could be convinced otherwise.)

> putting this info in the URL

This seems reasonable, especially because pagination layout seems like one of the things that you’d be less annoyed about being lost on reload—as opposed to something like the auto-reload flag.

One thing to note: the store module defined in this change assumes that all changes to the data go through it. That is, the only way to change the pagination limit is to call `setLimit`. It seems reasonable to assume that people won’t manually change `localStorage['TF.TensorBoard.PaginatedView.limit']`, or that if they do then they accept the consequences. However, the URL bar is, by design, user-manipulable state. If we store this info in the URL, then we have to watch for changes there, too, and that adds additional complexity.

Perhaps one step in the right direction would be to let `tf-storage` expose a function for creating a general-purpose listener, like `getInitializer` but Polymer-independent? Or something like that. I’m not quite sure.

> moving settings into the TensorBoard server

Don’t think I agree here. Each user should be able to set a different pagination limit, and change this without requiring frontend interaction. Semantically, too, it seems that the backend should tell the frontend what data to render, and not concern itself with how to render it. (Similarly, storing things like chart domains, x-axis values, etc. on the server seems wrong to me.)

Perhaps what you’re hoping for is some kind of “saved layout” that can be shared among users? For instance, a user sets up some settings and clicks “save,” and the settings are sent to the server; other users can then load the layout, which just changes their local settings.

> Mostly just bringing up future considerations.

Yep, sure thing.

**Comment by chihuahua** [comment]
How about when the need for another cross-plugin setting arises, we re-evaluate and see what we can refactor into a common module? I think that 2nd need will help inform us of what logic is common. As well as whether we want a server-stored setting, which would entail some complexity.
